libraries/chain/block_header_state.cpp

```diff
@@ -67,6 +67,48 @@ const vector<digest_type>& block_header_state::get_new_protocol_feature_activati
    return detail::get_new_protocol_feature_activations(header_exts);
 }
 
+finalizer_policy_diff block_header_state::calculate_finalizer_policy_diff(const finalizer_policy& new_policy) const {
+   if (finalizer_policies.empty()) {
+      return active_finalizer_policy->create_diff(new_policy);
+   }
+   for (const auto& e : finalizer_policies) {
+      if (e.second.state == finalizer_policy_tracker::state_t::pending) {
+         return e.second.policy->create_diff(new_policy);
+      }
+   }
+   for (const auto& e : finalizer_policies) {
+      if (e.second.state == finalizer_policy_tracker::state_t::proposed) {
+         return e.second.policy->create_diff(new_policy);
+      }
+   }
+   assert(false);
+}
+
+finalizer_policy block_header_state::calculate_finalizer_policy(const finalizer_policy_diff& diff) const {
+   finalizer_policy result;
+   if (finalizer_policies.empty()) {
+      assert(active_finalizer_policy);
+      result = *active_finalizer_policy;
+      result.apply_diff(diff);
+      return result;
+   }
+   for (const auto& e : finalizer_policies) {
+      if (e.second.state == finalizer_policy_tracker::state_t::pending) {
+         result = *e.second.policy;
+         result.apply_diff(diff);
+         return result;
+      }
+   }
+   for (const auto& e : finalizer_policies) {
+      if (e.second.state == finalizer_policy_tracker::state_t::proposed) {
+         result = *e.second.policy;
+         result.apply_diff(diff);
+         return result;
+      }
+   }
+   assert(false);
+}
```

Comment:
I believe these two functions don't do the right thing. They should return the **last** policy that will become active, not the the **first** one.

Also, both of them could use a utility function in `block_header_state`:

```c++
#include <boost/range/adaptor/reversed.hpp>

   const finalizer_policy_ptr& get_finalizer_policy_tail() const {
      if (!finalizer_policies.empty()) {
         auto reversed_finalizer_policies = boost::adaptors::reverse(finalizer_policies);
         for (const auto& e : reversed_finalizer_policies) {
            if (e.second.state == finalizer_policy_tracker::state_t::proposed)
               return e.second.policy;
         }
         return reversed_finalizer_policies.begin()->second.policy;
      }
      return active_finalizer_policy;
   }
```

Reply:
I completely agree. I don't understand how our existing tests didn't catch this. Need to add a new test case.

Reply:
> I don't understand how our existing tests didn't catch this. Need to add a new test case.

Because in the current finalizer_update_tests, we run as a proposer, and because we pass the original uncompressed policy to `finish_next()` from `next(block_header_state_input&)`. We don't `push_block()`.

So the compression/decompression is never activated in the test. If you extract the diff from the extension in `finish_next()` instead of taking the uncompressed policy, I think the tests would fail.

Reply:
The test is running a validator controller, so surprised that didn't catch it. I was looking into validating the diff, I think what you suggest is at least one way to do that.

Reply:
Yes, you are right it does a `push_block` to a `unique_ptr<controller> validating_node;`, but we don't check anything on that `validating_node` when it comes to finalizer_policy status.

Reply:
Calculating the policy instead of passing and the tests still pass. I can tell in the test that the if_ext has the wrong diff. It must be getting the right answer because the inverse of the diff is used with the same wrong value on both nodes. I've added a test that verifies the if_ext is correct.

Reply:
> Calculating ...  It must be getting the right answer because the inverse of the diff is used with the same wrong value on both nodes.

I don't think so. The issue is that we calculate a diff with the wrong source policy, but then we also use the same wrong source policy to apply the diff and restore the original `finalizer_policy`.

Reply:
I think we just said the same thing.

We do "compare" between both nodes because block ids have to match which includes the block header extensions.

Reply:
Oh, got it. Glad we agree. I understand your "compare" point now.